### PR TITLE
Exporter v2 with side inputs

### PR DESCRIPTION
**Pull request by kmindspark**
# Description

> :memo: Please include a summary of the change. 
>  
> * Please also include relevant motivation and context.  
> * List any dependencies that are required for this change.  

## Type of change

For a new feature or function, please create an issue first to discuss it
with us before submitting a pull request.

Note: Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation update
- [ ] TensorFlow 2 migration
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] A new research paper code implementation
- [ ] Other (Specify)

## Tests

> :memo: Please describe the tests that you ran to verify your changes.
>  
> * Provide instructions so we can reproduce.  
> * Please also list any relevant details for your test configuration.  

**Test Configuration**:

## Checklist

- [x] I have signed the [Contributor License Agreement](https://github.com/tensorflow/models/wiki/Contributor-License-Agreements).
- [x] I have read [guidelines for pull request](https://github.com/tensorflow/models/wiki/Submitting-a-pull-request).
- [x] My code follows the [coding guidelines](https://github.com/tensorflow/models/wiki/Coding-guidelines).
- [x] I have performed a self [code review](https://github.com/tensorflow/models/wiki/Code-review) of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

File: research/object_detection/exporter_lib_v2.py [hunk 1]
```diff
@@ -36,19 +36,41 @@ def _decode_tf_example(tf_example_string_tensor):
...
+    Args:
+      side_input_shapes: forward-slash-separated list of comma-separated lists
+        describing input shapes.
+      side_input_types: comma-separated list of the types of the inputs.
+      side_input_names: comma-separated list of the names of the inputs.
+
+    Returns:
+      a zipped list of side input tuples.
+    """
+    side_input_shapes = list(map(lambda x: eval('[' + x + ']'),
```
Comment (tombstone) [inline research/object_detection/exporter_lib_v2.py:53]: nit: can we use`ast.literal_eval`

File: research/object_detection/exporter_lib_v2.py [hunk 2]
```diff
@@ -75,11 +97,33 @@ def _run_inference_on_images(self, image):
...
+    sig = [tf.TensorSpec(shape=[1, None, None, 3], dtype=tf.uint8)]
+    if use_side_inputs:
+      for info in zipped_side_inputs:
+        self.side_input_names.append(info[2])
+        sig.append(tf.TensorSpec(shape=info[0],
+                                 dtype=info[1],
+                                 name=info[2]))
+
+    def __call__(input_tensor, *side_inputs):
+      kwargs = dict(zip(self.side_input_names.split(","), side_inputs))
```
Comment (tombstone) [inline research/object_detection/exporter_lib_v2.py:119]: as discussed offline, let's remove split.

Comment (beerys) [reply research/object_detection/exporter_lib_v2.py]: Just curious, what would you recommend is used instead?

File: research/object_detection/exporter_lib_v2.py [hunk 3]
```diff
@@ -75,11 +99,34 @@ def _run_inference_on_images(self, image):
...
+      zipped_side_inputs: the zipped side inputs.
+    """
+    self.side_input_names = []
+    sig = [tf.TensorSpec(shape=[1, None, None, 3], dtype=tf.uint8)]
+    if use_side_inputs:
+      for info in zipped_side_inputs:
+        self.side_input_names.append(info[2])
+        sig.append(tf.TensorSpec(shape=info[0],
+                                 dtype=info[1],
+                                 name=info[2]))
```
Comment (vighneshbirodkar) [inline research/object_detection/exporter_lib_v2.py:119]: I would suggest moving all of this logic into one function. Something like `parse_function_spec(shapes, types, names)`.

File: research/object_detection/exporter_lib_v2.py [hunk 4]
```diff
@@ -75,11 +99,34 @@ def _run_inference_on_images(self, image):
...
+
+    def call_func(self, input_tensor, *side_inputs):
+      kwargs = dict(zip(self.side_input_names, side_inputs))
+      return self._run_inference_on_images(input_tensor, **kwargs)
+
+    self.__call__ = tf.function(self.call_func, input_signature=sig)
```
Comment (vighneshbirodkar) [inline research/object_detection/exporter_lib_v2.py:125]: According to this
https://www.tensorflow.org/api_docs/python/tf/saved_model/save

function signatures are optional. Can you run a quick test to see what happens when you omit them?

File: research/object_detection/exporter_main_v2.py [hunk 5]
```diff
@@ -106,6 +106,27 @@
 flags.DEFINE_string('config_override', '',
                     'pipeline_pb2.TrainEvalPipelineConfig '
                     'text proto to override pipeline_config_path.')
+flags.DEFINE_boolean('use_side_inputs', False,
+                     'If True, uses side inputs as well as image inputs.')
+flags.DEFINE_string('side_input_shapes', "",
```
Comment (vighneshbirodkar) [inline research/object_detection/exporter_main_v2.py:111]: If it tuns out that these are required, I would suggest writing an example spec in the helper string.

File: research/object_detection/exporter_lib_v2.py (hunk 4 above)
Comment (vighneshbirodkar) [reply research/object_detection/exporter_lib_v2.py:125]: Another thought. If the user specifies tensor names from the corresponding dataset, can we infer shapes and types from there ?

Comment (kmindspark) [reply research/object_detection/exporter_lib_v2.py:125]: In order to create the saved model we need tensorspecs, doesn't work without it (at least within the limits of how we're diong it now). How would we infer shapes? Tensorspecs require shape as an argument so don't we need them at model-save-time? Also, the Tf1 exporter requires all three of these arguments so I sort of modeled after that. Would appreciate your thoughts.

Comment (vighneshbirodkar) [reply research/object_detection/exporter_lib_v2.py]: 1. We have access to the dataset, can we use that to infer the shape and types ? 
2. According to the documentation I linked above, function signatures are optional. If removing them (omitting the input signatures argument) gives an error, it world be good to have a record of it and clarify it with the tensorflow team for future reference.

Comment (kmindspark) [reply research/object_detection/exporter_lib_v2.py]: Talked with @tombstone, will leave this as a to-do.

**Comment by kmindspark** [comment]
Moved to new PR and merged (#9086)

**Pull request closed without merging**